### PR TITLE
Django bokeh

### PR DESCRIPTION
**Pull request by msouff**

### CODE REVIEW

File: tests/apps/tethysapp-test_app/tethysapp/test_app/controllers.py [hunk 1]
```diff
@@ -5,6 +5,16 @@
 from channels.generic.websocket import WebsocketConsumer
 import json
 
+import pandas as pd
```
[the remark this answers is not in the document]
Comment (sdc50) [reply tests/apps/tethysapp-test_app/tethysapp/test_app/controllers.py]: I guess we would need these as dependencies of the test app.

File: tethys_apps/base/app_base.py [hunk 2]
```diff
@@ -131,6 +136,82 @@ def url_patterns(self):
...
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
+                        try:
+                            handler_function = getattr(module, function_name)
+                        except AttributeError as e:
+                            error_msg = 'The following error occurred while trying to access the handler function ' \
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
```
Comment (swainn) [inline tethys_apps/base/app_base.py]: Lines 163-184 are virtually identical to lines 109-130. Please generalize into a  private method in this class. Something along the lines of `_resolve_function_ref(ref, ref_type)` used like `self._resolve_function_ref(url_map.controller, 'controller')`. The `ref_type` parameter would be used in the `error_msg`s.

File: tethys_apps/base/app_base.py [hunk 3]
```diff
@@ -131,6 +136,82 @@ def url_patterns(self):
...
+                            url_pattern = bokeh_app.url + suffix
+                            return f'^{url_pattern}$'
+
+                        http_url = url(urlpattern('/autoload.js'), BokehAutoloadJsCDN,
+                                       name=f'{url_map.name}_bokeh_autoload', kwargs=kwargs)
+                        ws_url = url(urlpattern('/ws'), WSConsumer, name=f'{url_map.name}_bokeh_ws', kwargs=kwargs)
+
+                        # Append to namespace list
+                        handler_patterns['http'][namespace].append(http_url)
+                        handler_patterns['websocket'][namespace].append(ws_url)
```
Comment (swainn) [inline tethys_apps/base/app_base.py]: I think we'd be better served if lines187-209 were extracted into a utility function/method as well.

File: tethys_apps/base/url_map.py [hunk 4]
```diff
@@ -17,7 +17,7 @@ class UrlMapBase:
 
     root_url = ''
 
-    def __init__(self, name, url, controller, protocol='http', regex=None):
+    def __init__(self, name, url, controller, protocol='http', regex=None, handler=None, handler_type=None):
         """
         Constructor
```
Comment (swainn) [inline tethys_apps/base/url_map.py]: This docstring needs to be updated with descriptions of the new `handler` and `handler_type` arguments. And `protocol` for that matter.

File: tethys_apps/base/url_map.py [hunk 5]
```diff
@@ -37,12 +37,15 @@ def __init__(self, name, url, controller, protocol='http', regex=None):
...
+        self.handler = handler
+        self.handler_type = handler_type
 
     def __repr__(self):
         """
         String representation
         """
-        return '<UrlMap: name={0}, url={1}, controller={2}>'.format(self.name, self.url, self.controller)
+        return f'<UrlMap: name={self.name}, url={self.url}, controller={self.controller}, protocol={self.protocol}, ' \
+            f'handler={self.handler}, handler_type={self.handler_type}>'
```
Comment (swainn) [inline tethys_apps/base/url_map.py]: I do love me some f strings. :D

File: tethys_portal/consumers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+"""
+********************************************************************************
+* Name: consumers/__init__.py
+* Author: Michael Souffront
+* Created On: 2019
+* Copyright: (c) Aquaveo, LLC 2019
```
Comment (swainn) [inline tethys_portal/consumers/__init__.py]: @sdc50 @SarvaPulla I'm not a legal expert, but what should the copyrights be in our Tethys Platform headers? I realized many of my contributions have Aquaveo listed as the copyright holder as well, b/c my editor adds it automatically.

File: tethys_portal/routing.py [hunk 7]
```diff
@@ -1,22 +1,46 @@
...
+for namespace, urls in app_http_handler_patterns.items():
+    for url in urls:
+        http_routing_patterns.append(url)
+
+for namespace, urls in app_ws_handler_patterns.items():
+    for url in urls:
+        ws_routing_patterns.append(url)
+
 application = ProtocolTypeRouter({
-    # Empty for now (http->django views is added by default)
```
Comment (swainn) [inline tethys_portal/routing.py]: Remove the "Empty for now" and parentheses portion of this comment. I think it is a good reminder to note that the view urls are added automatically.

File: tethys_portal/consumers/__init__.py (hunk 6 above)
Comment (SarvaPulla) [reply tethys_portal/consumers/__init__.py]: That's probably a good discussion for scrum/steering committee meeting. I believe Dan Ames might be able to help guide the discussion, since he has some experience in open source copyrights.

File: tethys_apps/base/app_base.py [hunk 8]
```diff
@@ -41,8 +45,59 @@ class TethysBase(TethysBaseMixin):
 
     def __init__(self):
         self._url_patterns = None
+        self._handler_patterns = None
         self._namespace = None
 
+    @staticmethod
+    def _resolve_ref_function(is_extension, ref, ref_type):
```
Comment (swainn) [inline tethys_apps/base/app_base.py]: @msouff please add docstrings to the new methods: `_resolve_ref_function` and `_resolve_bokeh_handler` using the Google style. Also, let's move `is_extension` to the back of the argument list.

File: tethys_portal/consumers/__init__.py (hunk 6 above)
Comment (swainn) [reply tethys_portal/consumers/__init__.py]: I haven't read this yet, but this series seems like it would be helpful: https://haacked.com/archive/2006/01/24/DevelopersGuideToOpenSourceSoftwareLicensing.aspx/

File: tethys_apps/base/url_map.py [hunk 9]
```diff
@@ -12,20 +12,23 @@
...
-    def __init__(self, name, url, controller, protocol='http', regex=None):
+    def __init__(self, name, url, controller, protocol='http', regex=None, handler=None, handler_type=None):
         """
         Constructor
 
         Args:
-          name (str): Name of the url map. Letters and underscores only (_).
-          url (str): Url pattern to map to the controller.
-          controller (str): Dot-notation path to the controller.
+          name (str): Name of the url map. Letters and underscores only (_). Most be unique within the app.
```
Comment (swainn) [inline tethys_apps/base/url_map.py]: Most -> Must

File: docs/tutorials/getting_started/bokeh_integration.rst [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+Create a new Tethys app named bokeh_tutorial.
+
+::
+
+    $ tethys scaffold bokeh_tutorial
+
+1. Bokeh Server
+===============
+
+``Bokeh`` is an interactive visualization library for Python. ``Bokeh Server`` is a component of the ``Bokeh`` architecture. It provides a way to sync model objects from Python to the client. This is done by levering the ``Websocket`` protocol. With the addition of ``Django Channels`` to Tethys, this ability to sync python objects and frontend plots has also been integrated without the need of other components such as a ``Tornado`` server (see `Tethys Bokeh Integration documentation <../../tethys_sdk/url_maps.html#bokeh-integration>`_). This integration facilitates the linking of objects and ``Bokeh`` widgets as well as the creation of the necessary ``websocket`` and ``http`` ``consumers``.
```
Comment (swainn) [inline docs/tutorials/getting_started/bokeh_integration.rst]: ... from Python to the client. -> ...in Python on the backend to JavaScript model objects on the client.

File: docs/tutorials/getting_started/bokeh_integration.rst [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+
+The logic for creating a Bokeh widget along with other related functionality is provided in a ``handler function``. This handler will be associated to a specific ``controller function`` where the resulting Bokeh widget will be displayed in a later step.
+
+2. Handler Functions Using Bokeh Widgets
+========================================
+
+Let's use Bokeh's sea temperature sample data to create a time series plot and link it to a slider that will provide the value to perform a rolling-window analysis on the time series. This a common example use in Bokeh's main documentation.
```
Comment (swainn) [inline docs/tutorials/getting_started/bokeh_integration.rst]: This a common example use in Bokeh's main documentation. -> This example is based on a similar example in Bokeh's main documentation.

File: docs/tutorials/getting_started/bokeh_integration.rst [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+
+a. Create a ``handler function`` by adding the following imports and logic to ``controller.py``.
```
Comment (swainn) [inline docs/tutorials/getting_started/bokeh_integration.rst]: Add the step to scaffold a new app. See the GeoServer tutorial. This can be simple like in the GeoServer tutorial: http://docs.tethysplatform.org/en/stable/tutorials/geoserver/start_and_register.html#scaffold-new-app

File: docs/tutorials/getting_started/bokeh_integration.rst [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+    ...
+
+    def home(request):
+        script = server_document(request.build_absolute_uri())
+        context = {'script': script}
+        return render(request, 'bokeh_tutorial/home.html', context)
+
+The home controller can now load the time series plot from (a) using the Bokeh ``server_document`` function. However, we still need to link the ``handler`` and the ``controller`` in the ``app.py``, and add the script context variable to the template as with any other variable.
+
+c. Modify ``app.py`` by adding a dot-formatter path to the handler function created in (a) to the ``handler`` parameter and providing a ``handler_type`` with a value equal to 'bokeh' as shown in the code below.
```
Comment (swainn) [inline docs/tutorials/getting_started/bokeh_integration.rst]: dot-formatter -> dot-formatted

File: docs/tutorials/getting_started/bokeh_integration.rst [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+.. figure:: ../../images/tutorial/bokeh_integration/bokeh_integration_2.png
+    :width: 650px
+
+The ``Slider`` and ``Plot`` will appear in the order they were added to the ``Column`` layout. If the value of the ``Slider`` changes, the data in the ``Plot`` will reflect this change based on this expression: `data = df.rolling(f'{new}D').mean()`. Where `df` is the sample data and `new` is the new ``Slider`` value.
+
+
+3. Handler Functions Using Param and Panel
+==========================================
+
+``Param`` is a Python library for providing parameters with dinamically generated values. One of the main advantages of ``Param`` is that parameters are provided using declarative programming. ``Panel``, on the other hand, is a visualization library for creating custom dashboards that rely on the use of widgets to render plots, images, and tables. These libraries can be used in combination with ``Bokeh Server`` to attain the same result of creating interactive tools within an app that are connected to Python objects. Given the depth of these libraries, the resulting code structure, and the level of difficulty for creating complex visualizations may be simplified.
```
Comment (swainn) [inline docs/tutorials/getting_started/bokeh_integration.rst]: dinamically -> dynamically

**Comment by swainn** [comment]
@msouff one last item. If you rebase with master again, I've just merged a PR that updates the What's New page for Tethys 3. Please add an entry to the What's New page with links to the new tutorial and API documentation.

**Comment by msouff** [comment]
This is ready

**Comment by swainn** [comment]
@sdc50 I have reviewed and approved this PR. Please take one final look and merge if you feel it is ready.

File: docs/whats_new.rst [hunk 15]
```diff
@@ -51,6 +51,22 @@ Django Channels
...
 See: :doc:`./tutorials/getting_started/websockets`
 
+Bokeh Integration
+-----------------
+
+* Bokeh's functionality to link backend Python objects to frontend JavaScript objects has been added.
+
+See: `Bokeh Integration Documentation <./tethys_sdk/url_maps.html#bokeh-integration>`_
+
+See: `Bokeh Integration Tutorial <./tutorials/getting_started/bokeh_integration.html>`_
```
Comment (swainn) [inline docs/whats_new.rst]: Do these links work? I'd prefer if you used the ref/doc link format, which is demonstrated many times in this document.

Comment (msouff) [reply docs/whats_new.rst]: If you look closely, the first link is referencing a sub section. I don't think you can do that with ref or doc, it'd have to be the whole document. That's why I did it like that, but I did test it.

I did change the second link to use :doc: and added a bokeh function that was missing from the mocked modules in the documentation.

Comment (swainn) [reply docs/whats_new.rst]: You can do a subsections with a ref, that's what they are for. You add the ref tag right before the subsection header and then use :ref: to reference it. There are many examples all throughout the documentation.

File: docs/conf.py [hunk 16]
```diff
@@ -26,7 +26,8 @@
 # NOTE: No obvious way to automatically anticipate all the sub modules without
 # installing the package, which is what we are trying to avoid.
 MOCK_MODULES = [
-    'bokeh', 'bokeh.embed', 'bokeh.resources',
```
Comment (swainn) [inline docs/conf.py]: Why did you remove these?

File: docs/whats_new.rst (hunk 15 above)
Comment (msouff) [reply docs/whats_new.rst]: Ahh, got it. To use ref from another file a tag needs to be created for the subsection. Done

File: docs/conf.py (hunk 16 above)
Comment (msouff) [reply docs/conf.py]: I didn't. I inserted a bunch of bokeh functions in there.
